### DISCUSSION TITLE
way to get i16 input even when it is not supported?

### DISCUSSION THREAD

**Discussion by droundy** [tomaka/cpal#1326] [Q&A]
There is code to convert from `f32` to `i16`, but I don't see how to write code that needs to work with `i16` in a portable way, that uses straight `i16` input where available, and otherwise performs a conversion.  Perhaps this already exists?

**Comment by LastExceed** [comment]
I would argue that this is beyond cpal's scope, as it is not something that needs to be abstracted across different backends, but rather something you would build on top of cpal

**Comment by roderickvd** [comment]
I'm not sure which "code to convert from `f32` to `i16`" you mean?

I used to agree with @LastExceed from the perspective of "cpal abstracts over system APIs". From the perspective of "cpal" as cross-platform abstraction however, this was discussed not so long ago [over at Discord](https://discord.com/channels/590254806208217089/672897096826748948/1525504566169436260):

> I guess we could add a `build_output_stream_converting::<T, D, E>()` that takes a `T` of the caller's choosing, and have cpal use `default_*_config()` internally to determine the device's default sample format and convert to that.

`dasp_sample` makes it easy so it wouldn't be hard to add or maintain, just a question of responsibility.

**Comment by LastExceed** [comment]
> I'm not sure which "code to convert from f32 to i16" you mean?

Probably the `FromSample` / `ToSample` trait impls
